src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+{
+    if (epoch < GROUP_KEY_SET_SIZE - 1)
+    {
+        return static_cast<uint8_t>(epoch + 1);
+    }
+    else
+    {
+        return 1;
+    }
+}
+
+static uint8_t previousEpoch(uint8_t epoch)
+{
+    if (epoch > 0)
+    {
+        return static_cast<uint8_t>(epoch - 1);
+    }
+    else
+    {
+        return static_cast<uint8_t>(GROUP_KEY_SET_SIZE - 1);
+    }
+}
+
+static EmberAfStatus printKeys()
+{
+    emberAfGroupKeyManagementClusterPrint("GROUP KEY SETS(%u)", sGroupKeys.size());
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it;
+    uint16_t fabricIndex = 0;
+    uint16_t keySet      = 0;
+
+    for (it = sGroupKeys.begin(); it != sGroupKeys.end(); ++it)
+    {
+        fabricIndex = std::get<0>(it->first);
+        keySet      = std::get<1>(it->first);
+
+        emberAfGroupKeyManagementClusterPrint("  fabric:%u, set:%u", fabricIndex, keySet);
+        uint8_t index     = it->second->current < GROUP_KEY_SET_SIZE ? it->second->current : 0;
+        RootKey & current = it->second->epoch[index];
+        if (current.active)
+        {
+            emberAfGroupKeyManagementClusterPrint("    current: key(%u), start(%u), policy:%u", sizeof(current.value),
+                                                  sizeof(current.startTime), current.policy);
+
+            index              = previousEpoch(index);
+            RootKey & previous = it->second->epoch[index];
+            if (previous.active)
+            {
+                emberAfGroupKeyManagementClusterPrint("    previous: key(%u), start(%u), policy:%u", sizeof(previous.value),
+                                                      sizeof(previous.startTime), previous.policy);
+
+                index         = previousEpoch(index);
+                RootKey & old = it->second->epoch[index];
+                if (old.active)
+                {
+                    emberAfGroupKeyManagementClusterPrint("    old: key(%u), start(%u), policy:%u", sizeof(old.value),
+                                                          sizeof(old.startTime), old.policy);
+                }
+            }
+        }
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static KeySet * getKeySet(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    return it == sGroupKeys.end() ? nullptr : it->second;
+}
+
+static EmberAfStatus setKey(uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan & keyRoot, chip::ByteSpan & epochStartTime,
+                            uint8_t securityPolicy)
+{
+    if (keyRoot.size() != GROUP_KEY_ROOT_SIZE || epochStartTime.size() != GROUP_KEY_EPOCH_SIZE)
+    {
+        return EMBER_ZCL_STATUS_INVALID_VALUE;
+    }
+    else
+    {
+        KeySet * set = getKeySet(fabricIndex, keySetIndex);
+        if (set)
+        {
+            set->current = nextEpoch(set->current);
```

Comment:
I don't think this works.  The spec distributes keys with GroupKeyIndex, which can be further divided into the 14-bit GroupKeySetIndex and the GroupKeyEpochIndex.

This PR instead conflates GroupKeyIndex and GroupKeySetIndex, and derives GroupKeyEpochIndex out of thin air simply based upon the time at which a key is set.  Out-of-order arrival of key set operations is absolutely possible.

So we need some restructuring around this.

* GroupKeyIndex needs to be explicitly decoded into GroupKeySetIndex and GroupKeyEpochIndex
* epoch index needs to come from the decoded value
* it may be advisable to verify that epoch indices are correctly ordered on GroupKeyStartTime, as the spec implies they should be